# Prelecture

In [3]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

print(df)

     row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
5        8     alice     Alice  female      koala     8-19      normal   
6        9      alli      Alli  female  alligator     11-8      snooty   
7       10    amelia    Amelia  female      eagle    11-19      snooty   
8       11  anabelle  Anabelle  female   anteater     2-16       peppy   
9       13   anchovy   Anchovy    male       bird      3-4        lazy   
10      14     angus     Angus    male       bull     4-30      cranky   
11      15  anicotti  Anicotti  female      mouse     2-24       peppy   
12      16     ankha     Ankha  female

1. Import the datatable and show it

In [4]:
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

1. Show the missing values in the dataset

In [5]:
df.shape

(391, 11)

2-1. Number of rows, Number of columns

2-2.

Observations: contain the data for each individual in a data set (ex: each row represents an observation)

Variables: the properties for each observation (ex: name, gender)

In [6]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

3. Generates summary like above, only works for numeric data types

In [7]:
df.describe(include = 'object')

id     name gender species birthday personality          song  \
count       390      391    391     391      391         391           380   
unique      390      391      2      35      361           8            92   
top     admiral  Admiral   male     cat     1-27        lazy  K.K. Country   
freq          1        1    204      23        2          60            10   

         phrase           full_id  \
count       391               391   
unique      388               391   
top     wee one  villager-admiral   
freq          2                 1   

                                                      url  
count                                                 391  
unique                                                391  
top     https://villagerdb.com/images/villagers/thumb/...  
freq                                                    1

3. Provides summary including non-numeric columns, non-null values, unique values

In [8]:
df['species'].value_counts()

species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

3. To get the frequency of unique values for a particular column

In [9]:
print(df.describe())

            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000


3. In jupyterhub, it will automatically display the result in a nicely formatted table without typing print()

4. My dataset has no missing values in numeric variables, so downloading another data below

In [10]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
sb = pd.read_csv(url)

pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

print(sb)

     survived  pclass     sex    age  sibsp  parch      fare embarked   class  \
0           0       3    male  22.00      1      0    7.2500        S   Third   
1           1       1  female  38.00      1      0   71.2833        C   First   
2           1       3  female  26.00      0      0    7.9250        S   Third   
3           1       1  female  35.00      1      0   53.1000        S   First   
4           0       3    male  35.00      0      0    8.0500        S   Third   
5           0       3    male    NaN      0      0    8.4583        Q   Third   
6           0       1    male  54.00      0      0   51.8625        S   First   
7           0       3    male   2.00      3      1   21.0750        S   Third   
8           1       3  female  27.00      0      2   11.1333        S   Third   
9           1       2  female  14.00      1      0   30.0708        C  Second   
10          1       3  female   4.00      1      1   16.7000        S   Third   
11          1       1  femal

In [7]:
sb

survived  pclass     sex    age  sibsp  parch      fare embarked   class  \
0           0       3    male  22.00      1      0    7.2500        S   Third   
1           1       1  female  38.00      1      0   71.2833        C   First   
2           1       3  female  26.00      0      0    7.9250        S   Third   
3           1       1  female  35.00      1      0   53.1000        S   First   
4           0       3    male  35.00      0      0    8.0500        S   Third   
5           0       3    male    NaN      0      0    8.4583        Q   Third   
6           0       1    male  54.00      0      0   51.8625        S   First   
7           0       3    male   2.00      3      1   21.0750        S   Third   
8           1       3  female  27.00      0      2   11.1333        S   Third   
9           1       2  female  14.00      1      0   30.0708        C  Second   
10          1       3  female   4.00      1      1   16.7000        S   Third   
11          1       1  female  58.00      0      0   26.5500        S   First   
12          0       3    male  20.00      0      0    8.0500        S   Third   
13          0       3    male  39.00      1      5   31.2750        S   Third   
14          0       3  female  14.00      0      0    7.8542        S   Third   
15          1       2  female  55.00      0      0   16.0000        S  Second   
16          0       3    male   2.00      4      1   29.1250        Q   Third   
17          1       2    male    NaN      0      0   13.0000        S  Second   
18          0       3  female  31.00      1      0   18.0000        S   Third   
19          1       3  female    NaN      0      0    7.2250        C   Third   
20          0       2    male  35.00      0      0   26.0000        S  Second   
21          1       2    male  34.00      0      0   13.0000        S  Second   
22          1       3  female  15.00      0      0    8.0292        Q   Third   
23          1       1    male  28.00      0      0   35.5000        S   First   
24          0       3  female   8.00      3      1   21.0750        S   Third   
25          1       3  female  38.00      1      5   31.3875        S   Third   
26          0       3    male    NaN      0      0    7.2250        C   Third   
27          0       1    male  19.00      3      2  263.0000        S   First   
28          1       3  female    NaN      0      0    7.8792        Q   Third   
29          0       3    male    NaN      0      0    7.8958        S   Third   
30          0       1    male  40.00      0      0   27.7208        C   First   
31          1       1  female    NaN      1      0  146.5208        C   First   
32          1       3  female    NaN      0      0    7.7500        Q   Third   
33          0       2    male  66.00      0      0   10.5000        S  Second   
34          0       1    male  28.00      1      0   82.1708        C   First   
35          0       1    male  42.00      1      0   52.0000        S   First   
36          1       3    male    NaN      0      0    7.2292        C   Third   
37          0       3    male  21.00      0      0    8.0500        S   Third   
38          0       3  female  18.00      2      0   18.0000        S   Third   
39          1       3  female  14.00      1      0   11.2417        C   Third   
40          0       3  female  40.00      1      0    9.4750        S   Third   
41          0       2  female  27.00      1      0   21.0000        S  Second   
42          0       3    male    NaN      0      0    7.8958        C   Third   
43          1       2  female   3.00      1      2   41.5792        C  Second   
44          1       3  female  19.00      0      0    7.8792        Q   Third   
45          0       3    male    NaN      0      0    8.0500        S   Third   
46          0       3    male    NaN      1      0   15.5000        Q   Third   
47          1       3  female    NaN      0      0    7.7500        Q   Third   
48          0       3    male    NaN

In [11]:
sb.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

There is both a) non-numeric variables and b) missing values in numeric variables in this dataset

In [12]:
sb.shape

(891, 15)

In [13]:
sb.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

4. The discrepancies between size of dataset given by 'sb.shape' and 'df.describe()':

a) number of columns it analyzes: 'sb.shape' shows the total number of columns while 'sb.describe()' only shows the numeric columns

b) the values it reports in the "count" column: 'sb.shape' shows the total number of rows while 'sb.describe()' represents the numbers of non-missing values in each column

""" the 'age' column is missing 177 rows so the number of 'age' when we typed sb.describe() is 891-177 = 714 """

In [14]:
df.describe

<bound method NDFrame.describe of      row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
5        8     alice     Alice  female      koala     8-19      normal   
6        9      alli      Alli  female  alligator     11-8      snooty   
7       10    amelia    Amelia  female      eagle    11-19      snooty   
8       11  anabelle  Anabelle  female   anteater     2-16       peppy   
9       13   anchovy   Anchovy    male       bird      3-4        lazy   
10      14     angus     Angus    male       bull     4-30      cranky   
11      15  anicotti  Anicotti  female      mouse     2-24       peppy   
12  

5. Difference between an "attribute" and a "method" :

Attributes are variables that belong to an object or class, representing the data.

Methods are functions that belong to an object or class; they define actions that an object can perform.

Ex.

df.describe -> This is a method object that doesn't execute but represents the method itself / outputs the reference

df.describe() -> This is a method call which output the dataframe countaining the summary statistics and executes the describe method

Chatgpt summary:

---

### Summary of Interaction

**Dataset Exploration and Analysis**

1. **Dataset Overview**:
   - **Dataset Used**: Characters from Animal Crossing, sourced from [here](https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv).
   - **Objective**: Explore and analyze the dataset to understand its structure, size, and the nature of missing values.

2. **Key Concepts Discussed**:
   - **Observations vs. Variables**:
     - **Observations**: Individual rows in the dataset, each representing a unique record or entry.
     - **Variables**: Columns in the dataset, each representing a different attribute or feature of the observations.

   - **Attributes vs. Methods**:
     - **Attributes**: Variables that store data about an object. For example, `my_car.color` retrieves the color of the car.
     - **Methods**: Functions that perform operations and return results. For example, `my_car.describe()` provides a description of the car.

3. **Pandas Functions**:
   - **`df.shape`**: Returns the dimensions of the DataFrame (number of rows and columns).
   - **`df.describe()`**: Provides summary statistics for numeric columns, including count, mean, standard deviation, min, and max. The "count" column shows the number of non-null values.
   - **`df.dropna()`**: Removes rows with missing values.
   - **`df.value_counts()`**: Counts unique values in a column and provides their frequency.

4. **Handling Missing Values**:
   - **`df.isnull().sum()`**: Shows the count of missing values per column.
   - **`df.dropna()`**: Used to remove rows with missing values.
   - **`del df['col']`**: Used to delete a column from the DataFrame.

5. **Displaying Data**:
   - **`pd.set_option('display.max_rows', None)`** and **`pd.set_option('display.max_columns', None)`**: Used to control the display settings of DataFrames in Pandas to show all rows and columns.

6. **Handling Errors**:
   - **`KeyError`**: Occurs when trying to access a column that does not exist.
   - **`AttributeError`**: Occurs when trying to use a method that does not exist on a DataFrame.

7. **Additional Concepts**:
   - **Size of the Dataset**: Refers to the number of rows and columns. `df.shape` gives these dimensions, while `df.describe()` provides counts of non-null values, which might differ due to missing data.

---

Chatgpt summary link: https://chatgpt.com/share/794e45cb-7746-4199-8559-8f3f239eff27

# Postlecture

In [15]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

6. Definitions of the summary statistics (can only be calculated for numeric variables in a dataset)

1) "count": the number of non-missing values
2) "mean": the average value of the column
3) "std": the standard deviation, measuring the spread and variability of the data
4) "min": the minimum value of the column
5) "25%": the 25th percentile, which is the value below which 25% of the data falls
6) "50%": median, the 50th percentil, which is the middle value
7) "75": the 3th quartile, the 75 percentile, which is the value below which 75% of the data falls
8) "max": the maximum value of the column

In [18]:
df_original = pd.read_csv(url)
df_copy = df_original.copy()
df_copy

survived  pclass     sex    age  sibsp  parch      fare embarked   class  \
0           0       3    male  22.00      1      0    7.2500        S   Third   
1           1       1  female  38.00      1      0   71.2833        C   First   
2           1       3  female  26.00      0      0    7.9250        S   Third   
3           1       1  female  35.00      1      0   53.1000        S   First   
4           0       3    male  35.00      0      0    8.0500        S   Third   
5           0       3    male    NaN      0      0    8.4583        Q   Third   
6           0       1    male  54.00      0      0   51.8625        S   First   
7           0       3    male   2.00      3      1   21.0750        S   Third   
8           1       3  female  27.00      0      2   11.1333        S   Third   
9           1       2  female  14.00      1      0   30.0708        C  Second   
10          1       3  female   4.00      1      1   16.7000        S   Third   
11          1       1  female  58.00      0      0   26.5500        S   First   
12          0       3    male  20.00      0      0    8.0500        S   Third   
13          0       3    male  39.00      1      5   31.2750        S   Third   
14          0       3  female  14.00      0      0    7.8542        S   Third   
15          1       2  female  55.00      0      0   16.0000        S  Second   
16          0       3    male   2.00      4      1   29.1250        Q   Third   
17          1       2    male    NaN      0      0   13.0000        S  Second   
18          0       3  female  31.00      1      0   18.0000        S   Third   
19          1       3  female    NaN      0      0    7.2250        C   Third   
20          0       2    male  35.00      0      0   26.0000        S  Second   
21          1       2    male  34.00      0      0   13.0000        S  Second   
22          1       3  female  15.00      0      0    8.0292        Q   Third   
23          1       1    male  28.00      0      0   35.5000        S   First   
24          0       3  female   8.00      3      1   21.0750        S   Third   
25          1       3  female  38.00      1      5   31.3875        S   Third   
26          0       3    male    NaN      0      0    7.2250        C   Third   
27          0       1    male  19.00      3      2  263.0000        S   First   
28          1       3  female    NaN      0      0    7.8792        Q   Third   
29          0       3    male    NaN      0      0    7.8958        S   Third   
30          0       1    male  40.00      0      0   27.7208        C   First   
31          1       1  female    NaN      1      0  146.5208        C   First   
32          1       3  female    NaN      0      0    7.7500        Q   Third   
33          0       2    male  66.00      0      0   10.5000        S  Second   
34          0       1    male  28.00      1      0   82.1708        C   First   
35          0       1    male  42.00      1      0   52.0000        S   First   
36          1       3    male    NaN      0      0    7.2292        C   Third   
37          0       3    male  21.00      0      0    8.0500        S   Third   
38          0       3  female  18.00      2      0   18.0000        S   Third   
39          1       3  female  14.00      1      0   11.2417        C   Third   
40          0       3  female  40.00      1      0    9.4750        S   Third   
41          0       2  female  27.00      1      0   21.0000        S  Second   
42          0       3    male    NaN      0      0    7.8958        C   Third   
43          1       2  female   3.00      1      2   41.5792        C  Second   
44          1       3  female  19.00      0      0    7.8792        Q   Third   
45          0       3    male    NaN      0      0    8.0500        S   Third   
46          0       3    male    NaN      1      0   15.5000        Q   Third   
47          1       3  female    NaN      0      0    7.7500        Q   Third   
48          0       3    male    NaN

(Running the latest url due to pd.read_csv(url)

In [22]:
df_copy.shape

(182, 15)

In [19]:
df_copy.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

Saving the Dataframe before making modifications.

The original copy is df_original. To modify, use df_copy.

In [20]:
df_copy.dropna(inplace=True)
df_copy

survived  pclass     sex    age  sibsp  parch      fare embarked   class  \
1           1       1  female  38.00      1      0   71.2833        C   First   
3           1       1  female  35.00      1      0   53.1000        S   First   
6           0       1    male  54.00      0      0   51.8625        S   First   
10          1       3  female   4.00      1      1   16.7000        S   Third   
11          1       1  female  58.00      0      0   26.5500        S   First   
21          1       2    male  34.00      0      0   13.0000        S  Second   
23          1       1    male  28.00      0      0   35.5000        S   First   
27          0       1    male  19.00      3      2  263.0000        S   First   
52          1       1  female  49.00      1      0   76.7292        C   First   
54          0       1    male  65.00      0      1   61.9792        C   First   
62          0       1    male  45.00      1      0   83.4750        S   First   
66          1       2  female  29.00      0      0   10.5000        S  Second   
75          0       3    male  25.00      0      0    7.6500        S   Third   
88          1       1  female  23.00      3      2  263.0000        S   First   
92          0       1    male  46.00      1      0   61.1750        S   First   
96          0       1    male  71.00      0      0   34.6542        C   First   
97          1       1    male  23.00      0      1   63.3583        C   First   
102         0       1    male  21.00      0      1   77.2875        S   First   
110         0       1    male  47.00      0      0   52.0000        S   First   
118         0       1    male  24.00      0      1  247.5208        C   First   
123         1       2  female  32.50      0      0   13.0000        S  Second   
124         0       1    male  54.00      0      1   77.2875        S   First   
136         1       1  female  19.00      0      2   26.2833        S   First   
137         0       1    male  37.00      1      0   53.1000        S   First   
139         0       1    male  24.00      0      0   79.2000        C   First   
148         0       2    male  36.50      0      2   26.0000        S  Second   
151         1       1  female  22.00      1      0   66.6000        S   First   
170         0       1    male  61.00      0      0   33.5000        S   First   
174         0       1    male  56.00      0      0   30.6958        C   First   
177         0       1  female  50.00      0      0   28.7125        C   First   
183         1       2    male   1.00      2      1   39.0000        S  Second   
193         1       2    male   3.00      1      1   26.0000        S  Second   
194         1       1  female  44.00      0      0   27.7208        C   First   
195         1       1  female  58.00      0      0  146.5208        C   First   
205         0       3  female   2.00      0      1   10.4625        S   Third   
209         1       1    male  40.00      0      0   31.0000        C   First   
215         1       1  female  31.00      1      0  113.2750        C   First   
218         1       1  female  32.00      0      0   76.2917        C   First   
224         1       1    male  38.00      1      0   90.0000        S   First   
230         1       1  female  35.00      1      0   83.4750        S   First   
245         0       1    male  44.00      2      0   90.0000        Q   First   
248         1       1    male  37.00      1      1   52.5542        S   First   
251         0       3  female  29.00      1      1   10.4625        S   Third   
252         0       1    male  62.00      0      0   26.5500        S   First   
257         1       1  female  30.00      0      0   86.5000        S   First   
262         0       1    male  52.00      1      1   79.6500        S   First   
263         0       1    male  40.00      0      0    0.0000        S   First   
268         1       1  female  58.00      0      1  153.4625        S   First   
269         1       1  female  35.00

In [21]:
df_copy.shape

(182, 15)

Now the rows with missing values are deleted.

In [23]:
del df_copy['fare']
df_copy

survived  pclass     sex    age  sibsp  parch embarked   class    who  \
1           1       1  female  38.00      1      0        C   First  woman   
3           1       1  female  35.00      1      0        S   First  woman   
6           0       1    male  54.00      0      0        S   First    man   
10          1       3  female   4.00      1      1        S   Third  child   
11          1       1  female  58.00      0      0        S   First  woman   
21          1       2    male  34.00      0      0        S  Second    man   
23          1       1    male  28.00      0      0        S   First    man   
27          0       1    male  19.00      3      2        S   First    man   
52          1       1  female  49.00      1      0        C   First  woman   
54          0       1    male  65.00      0      1        C   First    man   
62          0       1    male  45.00      1      0        S   First    man   
66          1       2  female  29.00      0      0        S  Second  woman   
75          0       3    male  25.00      0      0        S   Third    man   
88          1       1  female  23.00      3      2        S   First  woman   
92          0       1    male  46.00      1      0        S   First    man   
96          0       1    male  71.00      0      0        C   First    man   
97          1       1    male  23.00      0      1        C   First    man   
102         0       1    male  21.00      0      1        S   First    man   
110         0       1    male  47.00      0      0        S   First    man   
118         0       1    male  24.00      0      1        C   First    man   
123         1       2  female  32.50      0      0        S  Second  woman   
124         0       1    male  54.00      0      1        S   First    man   
136         1       1  female  19.00      0      2        S   First  woman   
137         0       1    male  37.00      1      0        S   First    man   
139         0       1    male  24.00      0      0        C   First    man   
148         0       2    male  36.50      0      2        S  Second    man   
151         1       1  female  22.00      1      0        S   First  woman   
170         0       1    male  61.00      0      0        S   First    man   
174         0       1    male  56.00      0      0        C   First    man   
177         0       1  female  50.00      0      0        C   First  woman   
183         1       2    male   1.00      2      1        S  Second  child   
193         1       2    male   3.00      1      1        S  Second  child   
194         1       1  female  44.00      0      0        C   First  woman   
195         1       1  female  58.00      0      0        C   First  woman   
205         0       3  female   2.00      0      1        S   Third  child   
209         1       1    male  40.00      0      0        C   First    man   
215         1       1  female  31.00      1      0        C   First  woman   
218         1       1  female  32.00      0      0        C   First  woman   
224         1       1    male  38.00      1      0        S   First    man   
230         1       1  female  35.00      1      0        S   First  woman   
245         0       1    male  44.00      2      0        Q   First    man   
248         1       1    male  37.00      1      1        S   First    man   
251         0       3  female  29.00      1      1        S   Third  woman   
252         0       1    male  62.00      0      0        S   First    man   
257         1       1  female  30.00      0      0        S   First  woman   
262         0       1    male  52.00      1      1        S   First    man   
263         0       1    male  40.00      0      0        S   First    man   
268         1       1  female  58.00      0      1        S   First  woman   
269         1       1  female  35.00      0      0        S   First  woman   
273         0       1    male  37.00      0      1        C   First    man   
275         1       1  fema

There's an error because 'id' has already been deleted since I have run it once.

In [24]:
df_copy.shape

(182, 14)

The column "id" is deleted.

In [25]:
df_original

survived  pclass     sex    age  sibsp  parch      fare embarked   class  \
0           0       3    male  22.00      1      0    7.2500        S   Third   
1           1       1  female  38.00      1      0   71.2833        C   First   
2           1       3  female  26.00      0      0    7.9250        S   Third   
3           1       1  female  35.00      1      0   53.1000        S   First   
4           0       3    male  35.00      0      0    8.0500        S   Third   
5           0       3    male    NaN      0      0    8.4583        Q   Third   
6           0       1    male  54.00      0      0   51.8625        S   First   
7           0       3    male   2.00      3      1   21.0750        S   Third   
8           1       3  female  27.00      0      2   11.1333        S   Third   
9           1       2  female  14.00      1      0   30.0708        C  Second   
10          1       3  female   4.00      1      1   16.7000        S   Third   
11          1       1  female  58.00      0      0   26.5500        S   First   
12          0       3    male  20.00      0      0    8.0500        S   Third   
13          0       3    male  39.00      1      5   31.2750        S   Third   
14          0       3  female  14.00      0      0    7.8542        S   Third   
15          1       2  female  55.00      0      0   16.0000        S  Second   
16          0       3    male   2.00      4      1   29.1250        Q   Third   
17          1       2    male    NaN      0      0   13.0000        S  Second   
18          0       3  female  31.00      1      0   18.0000        S   Third   
19          1       3  female    NaN      0      0    7.2250        C   Third   
20          0       2    male  35.00      0      0   26.0000        S  Second   
21          1       2    male  34.00      0      0   13.0000        S  Second   
22          1       3  female  15.00      0      0    8.0292        Q   Third   
23          1       1    male  28.00      0      0   35.5000        S   First   
24          0       3  female   8.00      3      1   21.0750        S   Third   
25          1       3  female  38.00      1      5   31.3875        S   Third   
26          0       3    male    NaN      0      0    7.2250        C   Third   
27          0       1    male  19.00      3      2  263.0000        S   First   
28          1       3  female    NaN      0      0    7.8792        Q   Third   
29          0       3    male    NaN      0      0    7.8958        S   Third   
30          0       1    male  40.00      0      0   27.7208        C   First   
31          1       1  female    NaN      1      0  146.5208        C   First   
32          1       3  female    NaN      0      0    7.7500        Q   Third   
33          0       2    male  66.00      0      0   10.5000        S  Second   
34          0       1    male  28.00      1      0   82.1708        C   First   
35          0       1    male  42.00      1      0   52.0000        S   First   
36          1       3    male    NaN      0      0    7.2292        C   Third   
37          0       3    male  21.00      0      0    8.0500        S   Third   
38          0       3  female  18.00      2      0   18.0000        S   Third   
39          1       3  female  14.00      1      0   11.2417        C   Third   
40          0       3  female  40.00      1      0    9.4750        S   Third   
41          0       2  female  27.00      1      0   21.0000        S  Second   
42          0       3    male    NaN      0      0    7.8958        C   Third   
43          1       2  female   3.00      1      2   41.5792        C  Second   
44          1       3  female  19.00      0      0    7.8792        Q   Third   
45          0       3    male    NaN      0      0    8.0500        S   Third   
46          0       3    male    NaN      1      0   15.5000        Q   Third   
47          1       3  female    NaN      0      0    7.7500        Q   Third   
48          0       3    male    NaN

In [26]:
df_original.shape

(891, 15)

7-1.

An example of 'use case' when df.dropna() might be preferred over using del df['col']

When we have a list of classmates but some of their information are missing because they are not coming, we use df.dropna() to clear all of the classmates who are not coming.

7-2.

An exmaple of 'use case' when del df['col'] might be preferred over using df.dropna()

When we want to delete whole column of information. If "age" is not important for employee information, then we can delete the whole column.

7-3.

The reason why doing del df['col'] first is because this makes the data cleaner and won't delete false data. For example, if the missing information is in the column we're deleting, deleting the column first change the row from missing info to not missing info.

7-4

Before: Before using df_copy.dropna() and del df_copy['col'], there are 391 rows and 11 columns. 

After: After using df_copy.dropna() and del df_copy['col'], there are 379 rows and 10 columns. Because 14 rows that have missing values were deleted by using df_copy.dropna() and 1 specific column was deleted by using del df_copy['col']

Chatgpy summary: 

---

**Summary of Interaction:**

In our discussion, we covered the following key points regarding handling and modifying DataFrames in Python using the `pandas` library:

1. **Saving and Restoring DataFrames:**
   - **Saving a DataFrame:** Use the `to_csv()` method to save a DataFrame to a CSV file for future reference.
   - **Loading a DataFrame:** Use `pd.read_csv()` to load a DataFrame from a CSV file.
   - **Creating a Backup:** Use `df.copy()` to create a copy of the DataFrame to preserve the original data before making modifications.

2. **Using `dropna()` Method:**
   - **Functionality:** The `dropna()` method is used to remove missing values from a DataFrame.
   - **In-Place Operation:** Use the `inplace=True` parameter if you want to modify the DataFrame in place, or assign the result back to the original DataFrame.

3. **Accessing Original Data:**
   - **Restoring Data:** If modifications are made, you can revert to the original DataFrame using the backup created earlier.

4. **Order of Operations:**
   - **Deleting Columns Before `dropna()`:** Removing unnecessary columns before applying `dropna()` can improve performance and reduce memory usage. It ensures that `dropna()` operates only on relevant columns and avoids unintended consequences.

---

Chatgpt chat log history link: https://chatgpt.com/share/ea724ff4-a5cc-4a93-8422-a27c70b0b7fb

In [12]:
import pandas as pd
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
dm = pd.read_csv(url)

pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

dm

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
30            31         0       1   
31            32         1       1   
32            33         1       3   
33            34         0       2   
34            35         0       1   
35            36         0       1   
36            37         1       3   
37            38         0       3   
38            39         0       3   
39            40         1       3   
40            41         0       3   
41            42         0       2   
42            43         0       3   
43            44         1       2   
44            45         1       3   
45            46         0       3   
46            47         0       3   
47            48         1       3   
48            49         0       3   
49            50         0       3   
50            51         0       3   
51            52         0       3   
52            53         1       1   
53            54         1       2   
54            55         0       1   
55            56         1       1   
56            57         1       2   
57            58         0       3   
58            59         1       2   
59            60         0       3   
60            61         0       3   
61            62         1       1   
62            63         0       1   
63            64         0       3   
64            65         0       1   
65            66         1       3   
66            67         1       2   
67            68         0       3   
68            69         1       3   
69            70         0       3   
70            71         0       2   
71            72         0       3   
72            73         0       2   
73            74         0       3   
74            75         1       3   
75            76         0       3   
76            77         0       3   
77            78         0       3   
78            79         1       2   
79            80         1       3   
80            81         0       3   
81            82         1       3   
82            83         1       3   
83            84         0       1   
84            85         1       2   
85            86         1       3   
86            87         0       3   
87            88         0       3   
88            89         1       1   
89            90         0       3   
90            91         0       3   
91            92         0       3   
92            93         0       1   
93            94         0       3   
94            95         0       3   
95            96         0       3   
96            97         0       1   
97            98         1       1   
98            99         1       2   
99           100         0       2   
100          101         0       3   
101          102         0       3   
102          103         0       1   
103          104         0       3   
104          10

Another dataset other than titanic.

In [32]:
dm.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [31]:
dm.groupby("Sex")["Cabin"].describe()

count unique top freq
Sex                         
female    97     75  G6    4
male     107     96  F2    3

There are 687 missing values in cabin, so the count value is 891-687 = 204, which is 97+107.

8-1.

'df.groupby("col1")["col2"].describe()' groups the Dataframe by the unique values in 'col1', and displays statistics for 'col2' for each group

For non-numeric column, there are count, unique, top, and freq.

1) count: the number of non-null entries
2) unique: the number of unique values in the column (how many different values)
3) top: the most frequent value (the mode)
4) freq: the frequency of the most frequent value (number of top)

In [13]:
dm.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

8-2.

Difference between 'count' value in df.describe() and df.groupby("col1")["col2"].describe() :

df.describe() -> The count value in here represents the number of non-null values in each column.

df.groupby("col1")["col2"].describe() -> The count value in here represents the number of non-null entries in col2 for each group of col1

'count' in the context of df.groupby is important for understanding the size and representativeness of each group within your data

In [1]:
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

print(df)

NameError: name 'pd' is not defined

8-3. i. 

Forget to include "import pandas as pd" in your code

In my opinion, ChatBot is definitely faster at troubleshoot and fixing the coding error than google search because it immediately tells you how to correct the code while google search may have various answers, or sometimes even an unspecified answer.

In [3]:
import pandas as pd
url = "titanic.csv"
sb = pd.read_csv(url)

pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

print(sb)

FileNotFoundError: [Errno 2] No such file or directory: 'titanic.csv'

In [4]:
import pandas as pd
url = "https://raws.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
sb = pd.read_csv(url)

pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

print(sb)

HTTPError: HTTP Error 500: Domain Not Found

In [6]:
import pandas as pd
url = "https://raw.githubusercontentt.com//mwaskom/seaborn-data/master/titanic.csv"
sb = pd.read_csv(url)

pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

print(sb)

URLError: <urlopen error [Errno -2] Name or service not known>

8-3. ii. Mistype "titanic.csv as "titanics.csv

There are different types of errors when making changes in the url (such as gaierror, HTTP Error, or URL Error)

In [13]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
5        8     alice     Alice  female      koala     8-19      normal   
6        9      alli      Alli  female  alligator     11-8      snooty   
7       10    amelia    Amelia  female      eagle    11-19      snooty   
8       11  anabelle  Anabelle  female   anteater     2-16       peppy   
9       13   anchovy   Anchovy    male       bird      3-4        lazy   
10      14     angus     Angus    male       bull     4-30      cranky   
11      15  anicotti  Anicotti  female      mouse     2-24       peppy   
12      16     ankha     Ankha  female        cat     9-22      snooty   
13      17  annalisa  Annalisa  female   anteater      2-6      normal   
14      18  annalise  Annalise  female      horse     12-2      snooty   
15      19   antonio   Antonio    male   anteater    10-20        jock   
16      20    apollo    Apollo    male      eagle      7-4      cranky   
17      21     apple     Apple  female    hamster     9-24       peppy   
18      22    astrid    Astrid  female   kangaroo      9-8      snooty   
19      23     audie     Audie  female       wolf     8-31       peppy   
20      24    aurora    Aurora  female    penguin     1-27      normal   
21      25       ava       Ava  female    chicken     4-28      normal   
22      26     avery     Avery    male      eagle     2-22      cranky   
23      27      axel      Axel    male   elephant     3-23        jock   
24      29   baabara   Baabara  female      sheep     3-28      snooty   
25      30       bam       Bam    male       deer     11-7        jock   
26      31    bangle    Bangle  female      tiger     8-27       peppy   
27      32    barold    Barold    male        cub      3-2        lazy   
28      33       bea       Bea  female        dog    10-15      normal   
29      34    beardo    Beardo    male       bear     9-27        smug   
30      35      beau      Beau    male       deer      4-5        lazy   
31      36     becky     Becky  female    chicken     12-9      snooty   
32      37     bella     Bella  female      mouse    12-28       peppy   
33      39  benedict  Benedict    male    chicken    10-10        lazy   
34      40  benjamin  Benjamin    male        dog      8-3        lazy   
35      41    bertha    Bertha  female      hippo     4-25      normal   
36      43   bettina   Bettina  female      mouse     6-12      normal   
37      45    bianca    Bianca  female      tiger    12-13       peppy   
38      46      biff      Biff    male      hippo     3-29        jock   
39      47   big-top   Big Top    male   elephant     10-3        lazy   
40      48      bill      Bill    male       duck      2-1        jock   
41      49     billy     Billy    male       goat     3-25        jock   
42      50    biskit    Biskit    male        dog     5-13        lazy   
43      51     bitty     Bitty  female      hippo     10-6      snooty   
44      52    blaire    Blaire  female   squirrel      7-3      snooty   
45      53   blanche   Blanche  female    ostrich    12-21      snooty   
46      55  bluebear  Bluebear  female        cub     6-24       peppy   
47      56       bob       Bob    male        cat      1-1        lazy   
48      57    bonbon    Bonbon  female     rabbit      3-3       peppy   
49      58     bones     Bones    male        dog      8-4        lazy   
50      59    boomer    Boomer    male    penguin      2-7        lazy   
51      60     boone     Boone    male    gorilla     9-12        jock   
52      61     boots     Boots    male  alligator      8-7        jock   
53      6

In [11]:
df.groupby('gender')['species'].describe()

count unique   top freq
gender                        
female   187     33   cat   14
male     204     34  frog   13

In [12]:
Df.groupby('gender')['species'].describe()

NameError: name 'Df' is not defined

8-3. iii. Try using a dataFrame before it's been assigned into the variable

DataFrame and variable names are Case-sensitive, which means df and Df are considered as different variables

In [14]:
df.describe(

SyntaxError: incomplete input (3431639942.py, line 1)

8-3. iv.

SyntaxError shows when python encounters incomplete code that isn't preperly closed or finished.

In [15]:
df.group_by('gender')['species'].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'

8-3. v.

AttibuteError occurs because pandas DataFrame does not have a method called group_by

In [18]:
df.groupby('Gender')['species'].describe()

KeyError: 'Gender'

In [19]:
df.groupby('gender')['Species'].describe()

KeyError: 'Column not found: Species'

8-3. vi.

KeyError occurs in pandas when you try to access a column that doesn't exist in the DataFrame

In [21]:
df.groupby(gender)['species'].describe()

NameError: name 'gender' is not defined

In [22]:
df.groupby('gender')[species].describe()

NameError: name 'species' is not defined

8-3. vii.

NameError occurs because we need to provide the column name as a string, not a bare variable. 

9. 

Yes, I have reviewed the course wiki-textbook and interacted with a CatBot to help me understand all the material in the tutorial and lecture that I didn't quite follow when I first saw it.

Chatgpt summary:

---

### Summary of Discussion

1. **DataFrame Analysis with `groupby()`**:
   - **Purpose**: We discussed the use of `df.groupby('column')` to group data in a DataFrame based on unique values in a column and then apply descriptive statistics on another column using `.describe()`.
   - **Code Example**: `df.groupby('gender')['species'].describe()` groups the DataFrame by the `'gender'` column and provides descriptive statistics for the `'species'` column.

2. **Error Handling in `pandas`**:
   - **`AttributeError`**: Occurs when calling a non-existent method (`df.group_by()` should be `df.groupby()`).
   - **`KeyError`**: Happens when referencing a column that does not exist in the DataFrame (e.g., `df.groupby('hand')` when `'hand'` is not a column).
   - **`NameError`**: Results from using an undefined name (e.g., `df.groupby(gender)` where `gender` is not defined; it should be `'gender'`).

3. **Case Sensitivity**:
   - **Python**: Variable names, method names, and column names are case-sensitive. Using incorrect case will result in errors.

4. **Column Names and Case Sensitivity**:
   - **Domain Names in URLs**: Not case-sensitive. Path and query parameters are case-sensitive.
   - **DataFrame Column Names**: Case-sensitive, so exact spelling and case must be used.

5. **Syntax and Code Completion**:
   - **SyntaxError**: Occurs when code is incomplete, such as missing closing parentheses.
   - **`SyntaxError: incomplete input`**: Indicates unfinished code that needs to be completed.

6. **Handling Errors**:
   - **Inspecting Columns**: Use `df.columns` to check available column names.
   - **Correcting Code**: Ensure correct method names (`groupby` instead of `group_by`) and proper use of column names as strings.

### Conclusion

Proper use of DataFrame methods and handling errors involves understanding case sensitivity, accurate column names, and correct syntax. Adjusting code based on the exact names and methods is crucial for successful data manipulation and analysis.

---

Chatgpt Chat Box history link: https://chatgpt.com/share/e2788ba0-1005-42c7-bce9-974b131ccd56